File: Medium/208. Implement Trie (Prefix Tree)/ImplementTrie.py
'''
LeetCode Medium: 208. Implement Trie (Prefix Tree)
https://leetcode.com/problems/implement-trie-prefix-tree/
'''
class TrieNode:

    def __init__(self):
        self.nxt = [None for i in range(26)]
        self.isEnd = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.root
        for i in word:
            if cur.nxt[ord(i)-ord('a')]==None:
                cur.nxt[ord(i)-ord('a')] = TrieNode()
            cur = cur.nxt[ord(i)-ord('a')]
        cur.isEnd = True

    def search(self, word: str) -> bool:
        cur = self.root
        for i in word:
            if cur.nxt[ord(i)-ord('a')]==None:
                return False
            cur = cur.nxt[ord(i)-ord('a')]
        return cur.isEnd

    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for i in prefix:
            if cur.nxt[ord(i)-ord('a')]==None:
                return False
            cur = cur.nxt[ord(i)-ord('a')]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: Medium/208. Implement Trie (Prefix Tree)/ImplementTrie.java
/*
LeetCode Medium: 208. Implement Trie (Prefix Tree)
https://leetcode.com/problems/implement-trie-prefix-tree/
*/
class TrieNode{
    public TrieNode[]nxt;
    public boolean isEnd;
    public TrieNode(){
        nxt = new TrieNode[26];
        isEnd = false;
    }
}

class Trie {
    public TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode cur = root;
        for(int i = 0;i<word.length();i++){
            if(cur.nxt[word.charAt(i)-'a']==null){
                cur.nxt[word.charAt(i)-'a'] = new TrieNode();
            }
            cur = cur.nxt[word.charAt(i)-'a'];
        }
        cur.isEnd = true;
    }
    
    public boolean search(String word) {
        TrieNode cur = root;
        for(int i = 0;i<word.length();i++){
            if(cur.nxt[word.charAt(i)-'a']==null){
                return false;
            }
            cur = cur.nxt[word.charAt(i)-'a'];
        }
        return cur.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for(int i = 0;i<prefix.length();i++){
            if(cur.nxt[prefix.charAt(i)-'a']==null){
                return false;
            }
            cur = cur.nxt[prefix.charAt(i)-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */